src/state.rs

```diff
@@ -167,30 +204,33 @@ impl NetworkState {
         A: Canon,
         R: Canon,
     {
-        let mut context = CallContext::new(self);
+        let mut snap = self.latest.clone();
+        let mut context = CallContext::new(&mut snap, self.modules.clone());
 
         let result =
             context.query(target, Query::from_canon(&query), gas_meter)?;
 
         result.cast().map_err(VMError::from_store_error)
     }
 
-    /// Transact with the contract at address `target`
+    /// Transact with the contract at `target` address in the latest snapshot,
+    /// returning both the result of the transaction and the snapshot ID of the
+    /// state before applying the transaction.
     pub fn transact<A, R>(
         &mut self,
         target: ContractId,
         transaction: A,
         gas_meter: &mut GasMeter,
-    ) -> Result<R, VMError>
+    ) -> Result<(R, [u8; 32]), VMError>
```

Comment:
This way, it looks like we are generating snapshots per transaction, when it should be per block.

The proposed API binds the future usage of this function to that behavior.

Even if internally we store the snapshots per transaction, this shouldn't be reflected in the API

Reply:
Based on what should a rollback be done? I agree that if snapshot is internal to rusk-vm, then we should not expose it.